### ISSUE TITLE
[REQ] Language specific configurable serializers

### ISSUE THREAD

**Issue by dotCipher** [OpenAPITools/openapi-generator#2418]
### Problem scope of this request

When generating code via the code generator process, the generators include specific serializer implementations as part of the generated code for handling (de)serialization of the models.

For example, if I am using the `java` generator, right now I automatically get the `Gson` serialization library as part of the generated code, however if I am using a client / server that already contains it's own serialization implementation (ie. `Jackson`), then I now need to have 2 serialization implementations on my classpath.

Another example I noticed when generating code via the `kotlin` generator was that `Moshi` was included by default (which only handles JSON to my knowledge?). However there are also other implementations of serializers that are a bit more flexible that exist (ie. [`kotlinx.serialzation`](https://github.com/Kotlin/kotlinx.serialization)).

## Proposed solution

I would be looking for a means to configure what serializer implementation to use based on the request / response types I have defined in my OpenAPI spec and the language of the generator I am using.  But have the serializer implementations be mapped to the the handled request / response types that might be defined in the OpenAPI spec being used.

For example, if I have an OpenAPI spec that has custom `schema` models being used as `application/json` and `application/xml` types, including the `Moshi` serialization implementation would not work when generating code to handle both request / response types.

As part of the code generation process, I would propose that the generator introspect the complete list of all request / response types (ie. `application/json` / `application/xml` / `application/protobuf` etc), **or** introspect what models are being mapped to what media types.  Then the generator should be able to have a lookup table of `Language -> Serializer_Implementation`, where each `Serializer_Implementation` would contain it's own `List<HandledMediaType>`.

The generator would then be able to pick the least number of `Serializer_Implementation`s that would fit serialization of all handled media types for that OpenAPI spec.  

Or on the flip side, the user would be able to configure the `Serializer_Implementation` they would like to use in their generated code, and the generator would write the appropriate code for that serializers implementation IFF that `Serializer_Implementation` chosen by the user actually handles the media types they have mapped in their OpenAPI spec.  For example if they configure the generator to use `Moshi` but have schema models mapped to `application/xml` media types, the code generation should fail fast and report an error.

## Alternatives I've considered

I might be naive here, but just wanted to flag this as something that wasn't obvious to me when dealing with serialization of the generated code models.  I have considered writing my own generator / template for the serializer implementations I would want, but unsure if that's the right path forward since it's high lift to spin up a new generator specifically for changing a few minor things based on the serializer implementations one would like to use.

**Comment by prat0088** [comment]
This feature would be useful to me, too, to support protobuf.  

What do the project owners think of @dotCipher's approach?  Is anything standing in the way of implementation?

**Comment by dotCipher** [comment]
Sadly this is something that is currently preventing me from using the openapi-generator, and instead I need to roll my own generation.

**Comment by dotCipher** [comment]
It's been over a year now, any news on this?

**Comment by rar91279** [comment]
> It's been over a year now, any opinions on this? Does my approach make sense?

Yes, it does IMHO. 
Resteasy client library generator for example is already configurable with Gson or Jackson per "serializationLibrary" parameter. 
Json-B and Moshi are 2 next libraries would be nice here. 
It make sense for my usecase too:
I would like to use OkHTTP client instead of Resteasy

**Comment by dotCipher** [comment]
Yeah sadly I've had to turn to other options instead of using open api for generation. This seems like a pretty major problem and it's a bit concerning that there has been no input around the solution here for years.
